File: cpp/benches/order_book_benches.cpp
#include "../src/order_book.h"
#include <benchmark/benchmark.h>
#include <random>
#include <vector>

static void BM_Add_10k_Orders(benchmark::State& state) {
    for (auto _ : state) {
        state.PauseTiming();
        OrderBook book;
        std::vector<Order> orders;
        orders.reserve(10000);
        for (int i = 0; i < 10000; ++i) {
            orders.emplace_back(i, Side::Buy, 100.0, 10);
        }
        state.ResumeTiming();

        for (auto& order : orders) {
            auto trades = book.add_order(order);
            benchmark::DoNotOptimize(trades);
        }
    }
}
BENCHMARK(BM_Add_10k_Orders);

static void BM_Cancel_1k_Orders(benchmark::State& state) {
    for (auto _ : state) {
        state.PauseTiming();
        OrderBook book;
        std::vector<Order> orders;
        std::mt19937 rng(42);
        std::uniform_int_distribution<int> dist(0, 9999);

        for (int i = 0; i < 10000; ++i) {
            Order order(i, Side::Buy, 100.0 + (i % 10), 10);
            orders.push_back(order);
            book.add_order(order);
        }

        std::vector<int> indices_to_cancel;
        indices_to_cancel.reserve(1000);
        for (int i = 0; i < 1000; ++i) {
            indices_to_cancel.push_back(dist(rng));
        }
        state.ResumeTiming();

        for (int idx : indices_to_cancel) {
            Order& order_to_cancel = orders[idx];
            book.cancel_order(order_to_cancel.order_id, order_to_cancel.side, order_to_cancel.price);
        }
    }
}
BENCHMARK(BM_Cancel_1k_Orders);

static void BM_Modify_1k_Orders(benchmark::State& state) {
    for (auto _ : state) {
        state.PauseTiming();
        OrderBook book;
        std::vector<Order> orders;
        std::mt19937 rng(42);
        std::uniform_int_distribution<int> dist(0, 9999);
        std::uniform_int_distribution<uint64_t> qty_dist(1, 200);

        for (int i = 0; i < 10000; ++i) {
            Order order(i, Side::Buy, 100.0 + (i % 10), 10);
            orders.push_back(order);
            book.add_order(order);
        }

        std::vector<int> indices_to_modify;
        indices_to_modify.reserve(1000);
        for (int i = 0; i < 1000; ++i) {
            indices_to_modify.push_back(dist(rng));
        }
        state.ResumeTiming();

        for (int idx : indices_to_modify) {
            Order& order_to_modify = orders[idx];
            book.modify_order(order_to_modify.order_id, order_to_modify.side, order_to_modify.price, qty_dist(rng));
        }
    }
}
BENCHMARK(BM_Modify_1k_Orders);

static void BM_MatchingEngine(benchmark::State& state) {
    for (auto _ : state) {
        state.PauseTiming();
        OrderBook book;
        int num_orders = state.range(0);
        for (int i = 0; i < num_orders; ++i) {
            Order order(i, Side::Sell, 100.0 + (i % 10), 10);
            book.add_order(order);
        }
        Order buy_order(num_orders, Side::Buy, 110.0, num_orders * 10);
        state.ResumeTiming();

        auto trades = book.add_order(buy_order);
        benchmark::DoNotOptimize(trades);
    }
}
BENCHMARK(BM_MatchingEngine)->Range(1000, 100000);

BENCHMARK_MAIN();

File: rust/benches/order_book_benches.rs
use criterion::{black_box, criterion_group, criterion_main, BenchmarkId, Criterion};
use lock_free_order_book::order::{Order, Side};
use lock_free_order_book::order_book::OrderBook;
use rand::prelude::*;

fn setup_book_with_orders(n: u32) -> (OrderBook, Vec<Order>) {
    let mut rng = StdRng::seed_from_u64(42);
    let mut book = OrderBook::new();
    let orders: Vec<Order> = (0..n)
        .map(|i| {
            let side = if rng.gen::<bool>() {
                Side::Buy
            } else {
                Side::Sell
            };
            let price = rng.gen_range(90..110);
            let quantity = rng.gen_range(1..100);
            Order::new(i as u64, side, price, quantity)
        })
        .collect();

    for order in &orders {
        book.add_order(black_box(*order));
    }

    (book, orders)
}

fn benchmark_add_order(c: &mut Criterion) {
    c.bench_function("add_10k_orders", |b| {
        b.iter(|| {
            let mut book = OrderBook::new();
            for i in 0..10_000 {
                book.add_order(black_box(Order::new(i, Side::Buy, 100, 10)));
            }
        })
    });
}

fn benchmark_cancel_order(c: &mut Criterion) {
    let (mut book, orders) = setup_book_with_orders(10_000);
    let mut rng = StdRng::seed_from_u64(42);

    c.bench_function("cancel_1k_orders", |b| {
        b.iter(|| {
            for _ in 0..1_000 {
                let order_to_cancel = orders.choose(&mut rng).unwrap();
                book.cancel_order(
                    black_box(order_to_cancel.order_id),
                    black_box(order_to_cancel.side),
                    black_box(order_to_cancel.price),
                );
            }
        })
    });
}

fn benchmark_modify_order(c: &mut Criterion) {
    let (mut book, orders) = setup_book_with_orders(10_000);
    let mut rng = StdRng::seed_from_u64(42);

    c.bench_function("modify_1k_orders", |b| {
        b.iter(|| {
            for _ in 0..1_000 {
                let order_to_modify = orders.choose(&mut rng).unwrap();
                let new_quantity = rng.gen_range(1..200);
                book.modify_order(
                    black_box(order_to_modify.order_id),
                    black_box(order_to_modify.side),
                    black_box(order_to_modify.price),
                    black_box(new_quantity),
                );
            }
        })
    });
}

fn benchmark_matching_engine(c: &mut Criterion) {
    let mut group = c.benchmark_group("matching_engine");
    for size in [1_000, 10_000, 100_000].iter() {
        group.bench_with_input(BenchmarkId::from_parameter(size), size, |b, &size| {
            b.iter_with_setup(
                || {
                    let mut book = OrderBook::new();
                    // Pre-fill the book with ask orders
                    for i in 0..size {
                        book.add_order(Order::new(i, Side::Sell, 100 + (i % 10), 10));
                    }
                    book
                },
                |mut book| {
                    // Send a sweeping buy order to match
                    book.add_order(black_box(Order::new(size, Side::Buy, 110, size * 10)));
                },
            );
        });
    }
    group.finish();
}

criterion_group!(
    benches,
    benchmark_add_order,
    benchmark_cancel_order,
    benchmark_modify_order,
    benchmark_matching_engine
);
criterion_main!(benches);
